balancer/leastrequest/balancer_test.go

```diff
@@ -0,0 +1,471 @@
+		t.Fatalf("grpc.Dial() failed: %v", err)
+	}
+	defer cc.Close()
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	testServiceClient := testgrpc.NewTestServiceClient(cc)
+
+	// Wait for all 3 backends to round robin across. The happens because a
+	// SubConn transitioning into READY causes a new picker update. Once the
+	// picker update with all 3 backends is present, this test can start to make
+	// assertions based on those backends.
+	if err := checkRoundRobinRPCs(ctx, testServiceClient, firstThreeAddresses); err != nil {
+		t.Fatalf("error in expected round robin: %v", err)
+	}
+
+	// Map ordering of READY SubConns is non deterministic. Thus, perform 3 RPCs
+	// mocked from the random to each index to learn the addresses of SubConns
+	// at each index.
+	index = 0
+	peerAtIndex := make([]string, 3)
+	var peer0 peer.Peer
+	if _, err := testServiceClient.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer0)); err != nil {
+		t.Fatalf("testServiceClient.EmptyCall failed: %v", err)
+	}
+	peerAtIndex[0] = peer0.Addr.String()
+	if _, err := testServiceClient.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer0)); err != nil {
+		t.Fatalf("testServiceClient.EmptyCall failed: %v", err)
+	}
+	peerAtIndex[1] = peer0.Addr.String()
+	if _, err := testServiceClient.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer0)); err != nil {
+		t.Fatalf("testServiceClient.EmptyCall failed: %v", err)
+	}
+	peerAtIndex[2] = peer0.Addr.String()
+
+	// Start streaming RPCs, but do not finish them. Each subsequent stream
+	// should be started according to the least request algorithm, and chosen
+	// between the indexes provided.
+	index = 0
+	indexes = []uint32{
+		0, 0, // Causes first stream to be on first address.
+		0, 1, // Compares first address (which already has a RPC) to second, so choose second.
+		1, 2, // Compares second address (which already has a RPC) to third, so choose third.
+	}
+	// Start a streaming call on first, but don't finish the stream.
+	var peer1 peer.Peer
+	stream1, err := testServiceClient.FullDuplexCall(ctx, grpc.Peer(&peer1))
+	if err != nil {
+		t.Fatalf("testServiceClient.FullDuplexCall failed: %v", err)
+	}
+
+	// Start a second streaming call. From the indexes injected into random
+	// number generator, this should compare Address 1, which already has a RPC
+	// to Address 2, so thus should start the stream on Address 2.
+	var peer2 peer.Peer
+	stream2, err := testServiceClient.FullDuplexCall(ctx, grpc.Peer(&peer2))
+	if err != nil {
+		t.Fatalf("testServiceClient.FullDuplexCall failed: %v", err)
+	}
+
+	// Start a third streaming call. From the indexes injected into random
+	// number generator, this should compare Address 2, which already has a RPC
+	// to Address 3, so thus should start the stream on Address 3.
+	var peer3 peer.Peer
+	stream3, err := testServiceClient.FullDuplexCall(ctx, grpc.Peer(&peer3))
+	if err != nil {
+		t.Fatalf("testServiceClient.FullDuplexCall failed: %v", err)
+	}
```

Comment:
Let's do these in a loop to avoid duplication:

```go
for range peerAtIndex {
  // Start streams & validate peer (using `peer.FromContext(stream.Context())`)
  // defer stream closure to keep it open
}
```

Reply:
Done. Thanks for the suggestion. Saved a lot of duplication :).